src/Compiler/Utilities/TypeHashing.fs

```diff
@@ -0,0 +1,345 @@
+
+[<AutoOpen>]
+module internal HashUtilities =
+
+    let private hashEntityRefName (xref: EntityRef) name =
+        let tag =
+            if xref.IsNamespace then
+                TextTag.Namespace
+            elif xref.IsModule then
+                TextTag.Module
+            elif xref.IsTypeAbbrev then
+                TextTag.Alias
+            elif xref.IsFSharpDelegateTycon then
+                TextTag.Delegate
+            elif xref.IsILEnumTycon || xref.IsFSharpEnumTycon then
+                TextTag.Enum
+            elif xref.IsStructOrEnumTycon then
+                TextTag.Struct
+            elif isInterfaceTyconRef xref then
+                TextTag.Interface
+            elif xref.IsUnionTycon then
+                TextTag.Union
+            elif xref.IsRecordTycon then
+                TextTag.Record
+            else
+                TextTag.Class
+
+        (hash tag) @@ (hashText name)
+
+    let hashTyconRefImpl (tcref: TyconRef) =
+        let demangled = tcref.DisplayNameWithStaticParameters
+        let tyconHash = hashEntityRefName tcref demangled
+
+        tcref.CompilationPath.AccessPath
+        |> hashListOrderMatters (fst >> hashText)
+        |> pipeToHash tyconHash
+
+module HashIL =
+
+    let hashILTypeRef (tref: ILTypeRef) =
+        tref.Enclosing
+        |> hashListOrderMatters hashText
+        |> addFullStructuralHash tref.Name
+
+    let private hashILArrayShape (sh: ILArrayShape) = sh.Rank
+
+    let rec hashILType (ty: ILType) : Hash =
+        match ty with
+        | ILType.Void -> hash ILType.Void
+        | ILType.Array(sh, t) -> hashILType t @@ hashILArrayShape sh
+        | ILType.Value t
+        | ILType.Boxed t -> hashILTypeRef t.TypeRef @@ (t.GenericArgs |> hashListOrderMatters (hashILType))
+        | ILType.Ptr t
+        | ILType.Byref t -> hashILType t
+        | ILType.FunctionPointer t -> hashILCallingSignature t
+        | ILType.TypeVar n -> hash n
+        | ILType.Modified(_, _, t) -> hashILType t
+
+    and hashILCallingSignature (signature: ILCallingSignature) =
+        let res = signature.ReturnType |> hashILType
+        signature.ArgTypes |> hashListOrderMatters (hashILType) |> pipeToHash res
+
+module HashAccessibility =
+
+    let isHiddenToObserver (TAccess access) (observer: ObserverVisibility) =
+        let isInternalCompPath x =
+            match x with
+            | CompPath(ILScopeRef.Local, _, []) -> true
+            | _ -> false
+
+        match access with
+        | [] -> false
+        | _ when List.forall isInternalCompPath access ->
+            match observer with
+            // The 'access' means internal, but our observer can see it (e.g. because of IVT attribute)
+            | PublicAndInternal -> false
+            | PublicOnly -> true
+        | _ -> true
+
+module rec HashTypes =
+    open Microsoft.FSharp.Core.LanguagePrimitives
+
+    let stampEquals g ty1 ty2 =
+        match (stripTyEqns g ty1), (stripTyEqns g ty2) with
```

Comment:
Refer to previous comment about stripping.

Reply:
I think it's fine here since it's a "utility" function. We don't usually assume in those if types were stripped or not. If we were to add new type, it should be done systematically across whole compiler.